mlkem/poly.c

```diff
@@ -224,34 +224,68 @@ void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES])
 
     #if (KYBER_POLYCOMPRESSEDBYTES == 128)
     for (i = 0; i < KYBER_N / 2; i++)
+        __CPROVER_assigns(i, __CPROVER_object_whole(r))
+        __CPROVER_loop_invariant(i >= 0)
+        __CPROVER_loop_invariant(i <= KYBER_N / 2)
+        __CPROVER_loop_invariant(__CPROVER_forall { unsigned int k; (k < i) ==>
+                                 ( r->coeffs[2 * k]     >= 0 && r->coeffs[2 * k]     < KYBER_Q &&
+                                   r->coeffs[2 * k + 1] >= 0 && r->coeffs[2 * k + 1] < KYBER_Q ) })
+        __CPROVER_decreases(KYBER_N / 2 - i)
     {
         // REF-CHANGE: Hoist scalar decompression into separate function
-        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[0] >> 0) & 0xF);
-        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[0] >> 4) & 0xF);
-        a += 1;
+        r->coeffs[2 * i + 0] = scalar_decompress_q_16((a[i] >> 0) & 0xF);
+        r->coeffs[2 * i + 1] = scalar_decompress_q_16((a[i] >> 4) & 0xF);
     }
+
+
+
     #elif (KYBER_POLYCOMPRESSEDBYTES == 160)
-    unsigned int j;
-    uint8_t t[8];
     for (i = 0; i < KYBER_N / 8; i++)
+        __CPROVER_assigns(i, __CPROVER_object_whole(r))
+        __CPROVER_loop_invariant(i >= 0)
+        __CPROVER_loop_invariant(i <= KYBER_N / 8)
```

Comment:
Should it be at line 246: `i < KYBER_N/8`?

Reply:
Afraid not... CBMC implicitly places the invariant just _before_ the final loop exit test, so I can indeed be equal to KYBER_N/8 just before the final loop exit.